src/miner.cpp

```diff
@@ -964,26 +965,33 @@ void BitcoinMiner(std::shared_ptr<CReserveScript> coinbaseScript, bool fProofOfS
         }
 
         CBlock *pblock = &pblocktemplate->block;
-        int32_t nNonceLocal;
 
         if (!fProofOfStake)
         {
             {
                 LOCK(cs_nonce);
-                nNonceLocal = nNonce_base++;
+                nExtraNonce = nNonce_base++;
                 if (!nTimeStart)
                     nTimeStart = GetTime();
             }
 
             pblock->nNonce = 0;
             pblock->nNonce64 = rand();
+            // Avoid holding cs_main during IncrementExtraNonce which doesn't require it.
+            unsigned int nHeight;
             {
                 LOCK(cs_main);
-                nExtraNonce = nNonceLocal;
-                IncrementExtraNonce(pblock, chainActive.Tip(), nExtraNonce);
+                nHeight = chainActive.Height();
             }
+            IncrementExtraNonce(pblock, nHeight, nExtraNonce);
```

Comment:
In this context, the cs_main lock is not needed, however the cs_main lock is actually needed for IncrementExtraNonce (or better yet a different lock)